common/src/main/java/org/apache/sedona/common/geography/Constructors.java

```diff
@@ -264,4 +264,271 @@ private static Geometry collectionToGeom(Geography g, GeometryFactory gf) {
+
+    builder.addPolyline(new S2Polyline(pts));
+
+    S2Error error = new S2Error();
+    if (!builder.build(error)) {
+      throw new IOException("Failed to build S2 polyline: " + error.text());
+    }
+    S2Polyline s2poly = layer.getPolyline();
+    return new SinglePolylineGeography(s2poly);
+  }
+
+  private static Geography mLineToGeog(Geometry geom) throws IOException {
+    MultiLineString mls = (MultiLineString) geom;
+    List<S2Polyline> features = new ArrayList<>();
+    for (int i = 0; i < mls.getNumGeometries(); i++) {
+      LineString ls = (LineString) mls.getGeometryN(i);
+      List<S2Point> pts = toS2Points(ls.getCoordinates());
+      if (pts.size() >= 2) {
+        S2Builder builder = new S2Builder.Builder().build();
+        S2PolylineLayer layer = new S2PolylineLayer();
+        builder.startLayer(layer);
+        builder.addPolyline(new S2Polyline(pts));
+        S2Error error = new S2Error();
+        if (!builder.build(error)) {
+          throw new IOException("Failed to build S2 polyline: " + error.text());
+        }
+        features.add(layer.getPolyline());
+      }
+      // Empty/degenerate parts are skipped
+    }
+    return new PolylineGeography(features);
+  }
+
+  private static Geography polyToGeog(Geometry geom) throws IOException {
+    // Build S2 loops: exterior + holes
+    Polygon poly = (Polygon) geom;
+    // Construct S2 polygon (parity handles holes automatically)
+    S2Polygon s2poly = toS2Polygon(poly);
+    PolygonGeography g = new PolygonGeography(s2poly);
+    g.setSRID(geom.getSRID());
+    return g;
+  }
+
+  private static Geography mPolyToGeog(Geometry geom) throws IOException {
+    final List<S2Polygon> polys = new ArrayList<>();
+
+    MultiPolygon mp = (MultiPolygon) geom;
+    for (int i = 0; i < mp.getNumGeometries(); i++) {
+      Polygon p = (Polygon) mp.getGeometryN(i);
+      S2Polygon s2 = toS2Polygon(p);
+      if (s2 != null && !s2.isEmpty()) polys.add(s2);
+    }
+
+    MultiPolygonGeography out =
+        new MultiPolygonGeography(
+            Geography.GeographyKind.MULTIPOLYGON, Collections.unmodifiableList(polys));
+    out.setSRID(geom.getSRID());
+    return out;
+  }
+
+  private static Geography geomCollToGeog(Geometry geom) throws IOException {
+    if (!(geom instanceof GeometryCollection)) {
+      throw new IllegalArgumentException(
+          "geomCollToGeog expects GeometryCollection, got " + geom.getGeometryType());
+    }
+    GeographyCollection coll = new GeographyCollection();
+
+    GeometryCollection gc = (GeometryCollection) geom;
+    for (int i = 0; i < gc.getNumGeometries(); i++) {
+      Geometry g = gc.getGeometryN(i);
+      Geography sub = null;
+
+      if (g instanceof Point) {
+        sub = pointToGeog((Point) g);
+      } else if (g instanceof MultiPoint) {
+        MultiPoint mp = (MultiPoint) g;
+        for (int k = 0; k < mp.getNumGeometries(); k++) {
+          coll.features.add(pointToGeog((Point) mp.getGeometryN(k)));
+        }
```

Comment:
This flattens multi point as separate geography objects instead of a single PointGeography. This does not look right.